narwhal/network/src/anemo_ext.rs

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use anemo::codegen::BoxError;
+use anemo::codegen::BoxFuture;
+use anemo::codegen::Service;
+use anemo::types::PeerEvent;
+use anemo::Network;
+use anemo::PeerId;
+use anemo::Request;
+use anemo::Response;
+use bytes::Bytes;
+use futures::FutureExt;
+
+pub trait NetworkExt {
+    fn waiting_peer(&self, peer_id: PeerId) -> WaitingPeer;
+}
+
+impl NetworkExt for Network {
+    fn waiting_peer(&self, peer_id: PeerId) -> WaitingPeer {
+        WaitingPeer::new(self.clone(), peer_id)
+    }
+}
+
+#[derive(Clone)]
+pub struct WaitingPeer {
+    peer_id: PeerId,
+    network: Network,
+}
+
+impl WaitingPeer {
+    pub fn new(network: Network, peer_id: PeerId) -> Self {
+        Self { peer_id, network }
+    }
+
+    async fn do_rpc(self, request: Request<Bytes>) -> Result<Response<Bytes>, BoxError> {
+        use tokio::sync::broadcast::error::RecvError;
+
+        let (mut subscriber, _) = self.network.subscribe();
```

Comment:
Is it worth moving this after the "already connected" check? Not sure how expensive this is

Reply:
The thinking here is that there could be a race between when we do the first "are we connected" check and when we subscribe to the events. This way we're guaranteed to see the event.